### PR TITLE
feature(zjow): add wandb logger features; fix relative bugs for wandb online logger

### PR DESCRIPTION
**Pull request by zjowowen**
## Description
Add and fix wandb logger features for rendering videos and logging information during training, which is tested in algorithms td3/ddpg/sac.
Fix relative bugs for wandb online logger.
Copy changes to wandb offline logger.

### CODE REVIEW

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/ddpg.py [hunk 1]
```diff
@@ -242,17 +242,24 @@ def _forward_learn(self, data: dict) -> Dict[str, Any]:
...
+            next_actor_data = self._target_model.forward(next_obs, mode='compute_actor')
+            next_actor_data['obs'] = next_obs
+            target_q_value = self._target_model.forward(next_actor_data, mode='compute_critic')['q_value']
+
         q_value_dict = {}
+        target_q_value_dict = {}
         if self._twin_critic:
             q_value_dict['q_value'] = q_value[0].mean()
             q_value_dict['q_value_twin'] = q_value[1].mean()
+            target_q_value_dict['target q_value'] = target_q_value[0].mean()
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/policy/ddpg.py:257]: fixed.

File: ding/bonus/ppof.py [hunk 2]
```diff
@@ -16,6 +17,11 @@
 from .config import get_instance_config, get_instance_env, get_hybrid_shape
 
 
+@dataclass
+class TrainingReturn:
+    wandb_url: str
+
```
Comment (zjowowen) [inline ding/bonus/ppof.py:23]: Add a data class return for catching wandb run urls when doing task initialization.

File: ding/bonus/ppof.py [hunk 3]
```diff
@@ -48,8 +54,11 @@ def __init__(
         if isinstance(env, str):
             assert env in PPOF.supported_env_list, "Please use supported envs: {}".format(PPOF.supported_env_list)
             self.env = get_instance_env(env)
-            assert cfg is None, 'It should be default env tuned config'
-            self.cfg = get_instance_config(env)
+            if cfg is None:
+                # 'It should be default env tuned config'
+                self.cfg = get_instance_config(env, algorithm="PPO")
+            else:
+                self.cfg = cfg
```
Comment (zjowowen) [inline ding/bonus/ppof.py:61]: Allow cfg to be loaded from outside files.

File: ding/bonus/ppof.py [hunk 4]
```diff
@@ -76,6 +85,10 @@ def __init__(
             )
         self.policy = PPOFPolicy(self.cfg, model=model)
 
+    def load_policy(self, policy_state_dict, config):
```
Comment (zjowowen) [inline ding/bonus/ppof.py:88]: Add API for loading policy state_dict as well as corresponding config.

File: ding/bonus/ppof.py [hunk 5]
```diff
@@ -85,24 +98,34 @@ def train(
...
             task.use(ppof_adv_estimator(self.policy))
             task.use(multistep_trainer(self.policy, log_freq=n_iter_log_show))
             task.use(CkptSaver(self.policy, save_dir=self.exp_name, train_freq=n_iter_save_ckpt))
-            task.use(wandb_online_logger(self.exp_name, metric_list=self.policy.monitor_vars(), anonymous=True))
+            task.use(
+                wandb_online_logger(
+                    self.exp_name,
+                    metric_list=self.policy.monitor_vars(),
+                    anonymous=True,
+                    project_name=self.exp_name,
```
Comment (zjowowen) [inline ding/bonus/ppof.py:120]: Add project_name for wandb logger, so as to assign a name of the wandb project.

File: ding/policy/ddpg.py [hunk 6]
```diff
@@ -453,3 +458,20 @@ def _monitor_vars_learn(self) -> List[str]:
         if self._twin_critic:
             ret += ['critic_twin_loss']
         return ret
+
+    def state_dict(self) -> Dict[str, Any]:
```
Comment (zjowowen) [inline ding/policy/ddpg.py:462]: Add state_dict and load_state_dict for policy to be called from outside.

File: ding/framework/middleware/functional/collector.py [hunk 7]
```diff
@@ -77,7 +78,7 @@ def _inference(ctx: "OnlineRLContext"):
 
         obs = {i: obs[i] for i in range(get_shape0(obs))}  # TBD
         inference_output = policy.forward(obs, **ctx.collect_kwargs)
-        ctx.action = [to_ndarray(v['action']) for v in inference_output.values()]  # TBD
+        ctx.action = np.array([to_ndarray(v['action']) for v in inference_output.values()])  # TBD
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/framework/middleware/functional/collector.py]: fixed.

File: ding/policy/ddpg.py [hunk 8]
```diff
@@ -314,6 +321,7 @@ def _forward_learn(self, data: dict) -> Dict[str, Any]:
             'td_error': td_error_per_sample.abs().mean(),
             **loss_dict,
             **q_value_dict,
+            **target_q_value_dict,
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/policy/ddpg.py:321]: fixed.

File: ding/framework/middleware/functional/logger.py [hunk 9]
```diff
@@ -147,7 +147,7 @@ def wandb_online_logger(
         return task.void()
```
[the remark this answers is not in the document]
Comment (zjowowen) [reply ding/framework/middleware/functional/logger.py:149]: fixed.

File: ding/framework/middleware/functional/collector.py [hunk 10]
```diff
@@ -1,6 +1,7 @@
 from typing import TYPE_CHECKING, Callable, List, Tuple, Any
 from easydict import EasyDict
 from functools import reduce
+import numpy as np
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/collector.py:4]: remove this

File: ding/framework/middleware/functional/evaluator.py [hunk 11]
```diff
@@ -257,7 +257,7 @@ def _evaluate(ctx: Union["OnlineRLContext", "OfflineRLContext"]):
                 eval_monitor.update_video(env.ready_imgs)
                 eval_monitor.update_output(inference_output)
             output = [v for v in inference_output.values()]
-            action = [to_ndarray(v['action']) for v in output]  # TBD
+            action = np.array([to_ndarray(v['action']) for v in output])  # TBD
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/evaluator.py:260]: the same problem

File: ding/bonus/config.py [hunk 12]
```diff
@@ -4,92 +4,157 @@
...
+                critic_head_layer_num=2,
+            )
+        else:
+            raise KeyError("not supported env type: {}".format(env))
+    elif algorithm == 'TD3':
+        cfg = TD3Policy.default_config()
+        if env == 'hopper':
+            cfg.update(
+                dict(
+                    exp_name='hopper_td3',
```
Comment (PaParaZz1) [inline ding/bonus/config.py:99]: simplify config, remove all the unmodifed fields

File: ding/bonus/ppof.py [hunk 13]
```diff
@@ -16,6 +17,11 @@
 from .config import get_instance_config, get_instance_env, get_hybrid_shape
 
 
+@dataclass
+class TrainingReturn:
+    wandb_url: str
```
Comment (PaParaZz1) [inline ding/bonus/ppof.py:22]: add related comments about this class

File: ding/policy/td3.py [hunk 14]
```diff
@@ -153,3 +154,7 @@ class from DDPG class by changing ``_actor_update_freq``, ``_twin_critic`` and n
             ),
         ),
     )
+
+    def monitor_vars(self) -> List[str]:
+        variables = ["q_value", "target q_value", "loss", "lr", "entropy", "target_q_value", "td_error"]
+        return variables
```
Comment (PaParaZz1) [inline ding/policy/td3.py:160]: directly return

File: ding/bonus/td3.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+from ding.data import DequeBuffer
+from ding.bonus.config import get_instance_config, get_instance_env, get_hybrid_shape
+
+
+@dataclass
+class TrainingReturn:
+    wandb_url: str
+
+
+class TD3:
```
Comment (PaParaZz1) [inline ding/bonus/td3.py:26]: rename to `OffPolicyAgent`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/framework/middleware/functional/collector.py (hunk 10 above)
Comment (zjowowen) [reply ding/framework/middleware/functional/collector.py:4]: fixed

File: ding/framework/middleware/functional/evaluator.py (hunk 11 above)
Comment (zjowowen) [reply ding/framework/middleware/functional/evaluator.py:260]: fixed

File: ding/bonus/config.py (hunk 12 above)
Comment (zjowowen) [reply ding/bonus/config.py:99]: fixed

File: ding/policy/td3.py (hunk 14 above)
Comment (zjowowen) [reply ding/policy/td3.py:160]: fixed

File: ding/bonus/td3.py (hunk 15 above)
Comment (zjowowen) [reply ding/bonus/td3.py:26]: fixed

File: ding/bonus/ppof.py (hunk 13 above)
Comment (zjowowen) [reply ding/bonus/ppof.py:27]: fixed

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**